operator/pkg/gateway-api/gateway_reconcile.go

```diff
@@ -276,11 +277,44 @@ func (r *gatewayReconciler) setAddressStatus(ctx context.Context, gw *gatewayv1b
 
 func (r *gatewayReconciler) setListenerStatus(ctx context.Context, gw *gatewayv1beta1.Gateway) error {
 	for _, l := range gw.Spec.Listeners {
-		cond := gatewayListenerReadyCondition(gw, true, "Listener Ready")
+		// SupportedKinds is a required field, so we can't declare it as nil.
+		supportedKinds := []gatewayv1beta1.RouteGroupKind{}
+		invalidRouteKinds := false
+		if l.AllowedRoutes != nil && len(l.AllowedRoutes.Kinds) != 0 {
+			for _, k := range l.AllowedRoutes.Kinds {
+				if groupDerefOr(k.Group, gatewayv1beta1.GroupName) == gatewayv1beta1.GroupName &&
+					k.Kind == getSupportedKind(l.Protocol) {
+					supportedKinds = append(supportedKinds, k)
+				} else {
+					invalidRouteKinds = true
+				}
+			}
+		} else {
+			supportedKinds = []gatewayv1beta1.RouteGroupKind{
+				{
+					Group: GroupPtr(gatewayv1beta1.GroupName),
+					Kind:  getSupportedKind(l.Protocol),
+				},
+			}
+		}
+		var conds []metav1.Condition
```

Comment:
We may need to update this with a special append, or move it to a hash, with the Type as the key, because Conditions, although they are stored as a slice, are actually interpreted by the apiserver as a map with cond.Type as the key. Using an intermediate hash, and then swapping the values back into a slice before applying the status update, should make sure this works as expected.

Reply:
good point, I have a merge utility methods, let me switch to it.

https://github.com/cilium/cilium/blob/d0a43aa05ee39b884ff9e87318e2a6aebd5ca98f/operator/pkg/gateway-api/status.go#L29